application/templates/secret.yaml

```
@@ -15,9 +15,24 @@ metadata:
   annotations:
 {{ toYaml $.Values.secret.annotations | indent 4 }}
 {{- end }}
+{{- if or .data .encodedData }}
 data:
+{{- if .data }}
 {{- range $key, $value := .data }}
   {{ $key }}: {{ $value | b64enc }}
 {{- end }}
 {{- end }}
+{{- if .encodedData }}
+{{- range $key, $value := .encodedData }}
+  {{ $key }}: {{ $value }}
+{{- end }}
+{{- end }}
+{{- if .stringData }}
+stringData:
+{{- range $key, $value := .stringData }}
+  {{ $key }}: {{ $value }}
+{{- end }}
+{{- end }}
+{{- end }}
+{{- end }}
```

It looks like this first `if` is closed at the wrong place: the way these conditionals are set up, data or string data are only included in this Secret if `.data` or `.encodedData` are truth-y. That said, I suggest avoiding conditionals altogether:

```suggestion
data:
  {{- range $key, $value := .data }}
  {{ $key }}: {{ $value | b64enc }}
  {{- end }}
  {{- range $key, $value := .encodedData }}
  {{ $key }}: {{ $value }}
  {{- end }}
stringData:
  {{- toYaml .stringData | nindent 2 }}
```

This way, `data` will turn `null` due to a completely empty value in the rendered template if neither `.data` nor `.encodedData` are populated, and `stringData` will be an explicit `null` if `.stringData` is itself a null value.

What do you think? I think the unit test you wrote will still work.

---

Nice change. Smart one! Great!